In [1]:
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt 
import seaborn as sns
data = pd.read_csv("employee_promotion.csv")
data.head()

employee_id         department     region         education gender  \
0        65438  Sales & Marketing   region_7  Master's & above      f   
1        65141         Operations  region_22        Bachelor's      m   
2         7513  Sales & Marketing  region_19        Bachelor's      m   
3         2542  Sales & Marketing  region_23        Bachelor's      m   
4        48945         Technology  region_26        Bachelor's      m   

  recruitment_channel  no_of_trainings  age  previous_year_rating  \
0            sourcing                1   35                   5.0   
1               other                1   30                   5.0   
2            sourcing                1   34                   3.0   
3               other                2   39                   1.0   
4               other                1   45                   3.0   

   length_of_service  KPIs_met >80%  awards_won?  avg_training_score  \
0                  8              1            0                  49   
1                  4              0            0                  60   
2                  7              0            0                  50   
3                 10              0            0                  50   
4                  2              0            0                  73   

   is_promoted  
0            0  
1            0  
2            0  
3            0  
4            0

In [2]:
# Class Distrubtion 分佈
target_class = data.is_promoted.value_counts()
print('Class 0:', target_class[0])
print('Class 1:', target_class[1])
print('Ratio:', round(target_class[0] / target_class[1], 2), ': 1')

Class 0: 50140
Class 1: 4668
Ratio: 10.74 : 1


In [3]:
data.isnull().sum()

employee_id                0
department                 0
region                     0
education               2409
gender                     0
recruitment_channel        0
no_of_trainings            0
age                        0
previous_year_rating    4124
length_of_service          0
KPIs_met >80%              0
awards_won?                0
avg_training_score         0
is_promoted                0
dtype: int64

In [4]:
# Imputing the Missing values
data['education'].fillna(data['education'].mode()[0], inplace = True)
data['previous_year_rating'].fillna(data['previous_year_rating'].mode()[0], inplace=True)
data.isnull().sum()

employee_id             0
department              0
region                  0
education               0
gender                  0
recruitment_channel     0
no_of_trainings         0
age                     0
previous_year_rating    0
length_of_service       0
KPIs_met >80%           0
awards_won?             0
avg_training_score      0
is_promoted             0
dtype: int64

In [5]:
from sklearn.preprocessing import LabelEncoder
lb = LabelEncoder()
data['department'] = lb.fit_transform(data['department'])
data['region'] = lb.fit_transform(data['region'])
data['education'] = lb.fit_transform(data['education'])
data['gender'] = lb.fit_transform(data['gender'])
data['recruitment_channel'] = lb.fit_transform(data['recruitment_channel'])
data.select_dtypes('object').columns

Index([], dtype='object')

In [6]:
# Let's split the Target Data from the Dataset
X=data.iloc[:,:-1]
y=data.iloc[:,-1]
print("The Shape of X: ", X.shape)
print("The Shape of y: ", y.shape)

The Shape of X:  (54808, 13)
The Shape of y:  (54808,)


In [7]:
# Split DB into Training and Testing Sets

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)
print("The Shape of X train :", X_train.shape)
print("The Shape of X test :", X_test.shape)
print("The Shape of Y train :", y_train.shape)
print("The Shape of Y test :", y_test.shape)

The Shape of X train : (43846, 13)
The Shape of X test : (10962, 13)
The Shape of Y train : (43846,)
The Shape of Y test : (10962,)


## 【 Machine Learning Models】
### Logistic Regression 羅吉斯迴歸

In [8]:
# Lets Build a Predictive Model using the Random Forest Classifier

from sklearn.linear_model import LogisticRegression
model1 = LogisticRegression()
model1.fit(X_train,y_train)
y_pred1 = model1.predict(X_test)

In [9]:
from sklearn.metrics import classification_report, accuracy_score
print(accuracy_score(y_test, y_pred1))
print(classification_report(y_test, y_pred1))

0.9159824849480022
              precision    recall  f1-score   support

           0       0.92      1.00      0.96     10041
           1       0.00      0.00      0.00       921

    accuracy                           0.92     10962
   macro avg       0.46      0.50      0.48     10962
weighted avg       0.84      0.92      0.88     10962



C:\Users\cti110016\Anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\cti110016\Anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\cti110016\Anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


###  Random Forest 隨機森林

In [10]:
# Lets Build a Predictive Model using the Random Forest Classifier

from sklearn.ensemble import RandomForestClassifier
model = RandomForestClassifier()
model.fit(X_train,y_train)
y_pred = model.predict(X_test)

In [11]:
from sklearn.metrics import classification_report, accuracy_score
print(accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))

0.9347746761539865
              precision    recall  f1-score   support

           0       0.94      1.00      0.97     10041
           1       0.90      0.25      0.39       921

    accuracy                           0.93     10962
   macro avg       0.92      0.62      0.68     10962
weighted avg       0.93      0.93      0.92     10962



## 【Random Sampling 隨機採樣法】
### Random OverSampling 隨機過採

In [12]:
from imblearn.over_sampling import RandomOverSampler
ros = RandomOverSampler(random_state=0)
ros.fit(X, y)
X_resampled, y_resampled = ros.fit_resample(X, y)
y_resampled.value_counts()

1    50140
0    50140
Name: is_promoted, dtype: int64

In [13]:
# Splitting the dataset
X_train_re, X_test_re, y_train_re, y_test_re =train_test_split(X_resampled, y_resampled,
                                                               test_size=0.2, random_state=0)
print("The Shape of X train :", X_train_re.shape)
print("The Shape of X test :", X_test_re.shape)
print("The Shape of Y train :", y_train_re.shape)
print("The Shape of Y test :", y_test_re.shape)

The Shape of X train : (80224, 13)
The Shape of X test : (20056, 13)
The Shape of Y train : (80224,)
The Shape of Y test : (20056,)


In [14]:
# Lets Build a Predictive Model using the Random Forest Classifier

from sklearn.ensemble import RandomForestClassifier
model2 = RandomForestClassifier()
model2.fit(X_train_re, y_train_re)
y_pred_re = model2.predict(X_test_re) # Predictions for the X test

In [15]:
# Accuracy score and classification report

from sklearn.metrics import classification_report, accuracy_score
print(accuracy_score(y_test_re, y_pred_re))
print(classification_report(y_test_re, y_pred_re))

0.9920223374551257
              precision    recall  f1-score   support

           0       1.00      0.98      0.99      9981
           1       0.98      1.00      0.99     10075

    accuracy                           0.99     20056
   macro avg       0.99      0.99      0.99     20056
weighted avg       0.99      0.99      0.99     20056



### Random Undersampling 隨機欠採樣 

In [16]:
from imblearn.under_sampling import RandomUnderSampler
ros = RandomUnderSampler(random_state=0)
ros.fit(X, y)
X_resampled2, y_resampled2 = ros.fit_resample(X, y) # Making the Samples

In [17]:
# lets check the Class Distribution
y_resampled2.value_counts()

1    4668
0    4668
Name: is_promoted, dtype: int64

In [18]:
# Splitting the dataset
X_train_us, X_test_us, y_train_us, y_test_us =train_test_split(X_resampled2, y_resampled2,
                                                               test_size=0.2, random_state=0)
print("The Shape of X train :", X_train_us.shape)
print("The Shape of X test :", X_test_us.shape)
print("The Shape of Y train :", y_train_us.shape)
print("The Shape of Y test :", y_test_us.shape)

The Shape of X train : (7468, 13)
The Shape of X test : (1868, 13)
The Shape of Y train : (7468,)
The Shape of Y test : (1868,)


In [19]:
# Lets Build a Predictive Model using the Random Forest Classifier

from sklearn.ensemble import RandomForestClassifier
model3 = RandomForestClassifier()
model3.fit(X_train_us, y_train_us)
y_pred_us = model3.predict(X_test_us) # Predictions for the X test

In [20]:
# Accuracy score and classification report

from sklearn.metrics import classification_report, accuracy_score
print(accuracy_score(y_test_us, y_pred_us))
print(classification_report(y_test_us, y_pred_us))

0.7906852248394004
              precision    recall  f1-score   support

           0       0.85      0.71      0.77       929
           1       0.75      0.87      0.81       939

    accuracy                           0.79      1868
   macro avg       0.80      0.79      0.79      1868
weighted avg       0.80      0.79      0.79      1868



### Synthetic Sampling (SMOTE) 相似合成採樣

In [21]:
from imblearn.over_sampling import SMOTE
sm = SMOTE(random_state=0)
sm.fit(X, y)
X_resampled3, y_resampled3 = sm.fit_resample(X, y) # Making the Samples

In [22]:
# lets check the Class Distribution
y_resampled3.value_counts()

1    50140
0    50140
Name: is_promoted, dtype: int64

In [23]:
# Splitting the dataset
X_train_sm, X_test_sm, y_train_sm, y_test_sm =train_test_split(X_resampled3, y_resampled3,
                                                               test_size=0.2, random_state=0)
print("The Shape of X train :", X_train_sm.shape)
print("The Shape of X test :", X_test_sm.shape)
print("The Shape of Y train :", y_train_sm.shape)
print("The Shape of Y test :", y_test_sm.shape)

The Shape of X train : (80224, 13)
The Shape of X test : (20056, 13)
The Shape of Y train : (80224,)
The Shape of Y test : (20056,)


In [24]:
# Lets Build a Predictive Model using the Random Forest Classifier

from sklearn.ensemble import RandomForestClassifier
model4 = RandomForestClassifier()
model4.fit(X_train_sm, y_train_sm)
y_pred_sm = model4.predict(X_test_sm) # Predictions for the X test

In [25]:
# Accuracy score and classification report

from sklearn.metrics import classification_report, accuracy_score
print(accuracy_score(y_test_sm, y_pred_sm))
print(classification_report(y_test_sm, y_pred_sm))

0.9474471479856402
              precision    recall  f1-score   support

           0       0.93      0.97      0.95      9981
           1       0.97      0.93      0.95     10075

    accuracy                           0.95     20056
   macro avg       0.95      0.95      0.95     20056
weighted avg       0.95      0.95      0.95     20056



## Using Neighbors Based Sampling

In [26]:
from imblearn.under_sampling import EditedNearestNeighbours
enn = EditedNearestNeighbours()
enn.fit(X, y)
X_resampled4, y_resampled4 = enn.fit_resample(X, y) # Making the Samples

In [27]:
# lets check the Class Distribution
y_resampled4.value_counts()

0    39220
1     4668
Name: is_promoted, dtype: int64

In [28]:
# Splitting the dataset
X_train_enn, X_test_enn, y_train_enn, y_test_enn =train_test_split(X_resampled4, y_resampled4,
                                                               test_size=0.2, random_state=0)
print("The Shape of X train :", X_train_enn.shape)
print("The Shape of X test :", X_test_enn.shape)
print("The Shape of Y train :", y_train_enn.shape)
print("The Shape of Y test :", y_test_enn.shape)

The Shape of X train : (35110, 13)
The Shape of X test : (8778, 13)
The Shape of Y train : (35110,)
The Shape of Y test : (8778,)


In [29]:
# Lets Build a Predictive Model using the Random Forest Classifier

from sklearn.ensemble import RandomForestClassifier
model5 = RandomForestClassifier()
model5.fit(X_train_enn, y_train_enn)
y_pred_enn = model5.predict(X_test_enn)

In [30]:
# Accuracy score and classification report

from sklearn.metrics import classification_report, accuracy_score
print(accuracy_score(y_test_enn, y_pred_enn))
print(classification_report(y_test_enn, y_pred_enn))

0.9141034404192299
              precision    recall  f1-score   support

           0       0.92      0.99      0.95      7785
           1       0.86      0.29      0.43       993

    accuracy                           0.91      8778
   macro avg       0.89      0.64      0.69      8778
weighted avg       0.91      0.91      0.89      8778



## Combination of Over and UnderSampling

In [31]:
from imblearn.combine import SMOTEENN
sn = SMOTEENN(random_state = 0)
sn.fit(X, y)
X_resampled5, y_resampled5 = sn.fit_resample(X, y) # Making the Samples

In [32]:
# lets check the Class Distribution
y_resampled5.value_counts()

1    42331
0    26832
Name: is_promoted, dtype: int64

In [33]:
# Splitting the dataset
X_train_sn, X_test_sn, y_train_sn, y_test_sn =train_test_split(X_resampled5, y_resampled5,
                                                               test_size=0.2, random_state=0)

print("The Shape of X train :", X_train_sn.shape)
print("The Shape of X test :", X_test_sn.shape)
print("The Shape of Y train :", y_train_sn.shape)
print("The Shape of Y test :", y_test_sn.shape)

The Shape of X train : (55330, 13)
The Shape of X test : (13833, 13)
The Shape of Y train : (55330,)
The Shape of Y test : (13833,)


In [34]:
# Lets Build a Predictive Model using the Random Forest Classifier

from sklearn.ensemble import RandomForestClassifier
model6 = RandomForestClassifier()
model6.fit(X_train_sn, y_train_sn)
y_pred_sn = model6.predict(X_test_sn)

In [35]:
# Accuracy score and classification report

from sklearn.metrics import classification_report, accuracy_score
print(accuracy_score(y_test_sn, y_pred_sn))
print(classification_report(y_test_sn, y_pred_sn))

0.9383358635147835
              precision    recall  f1-score   support

           0       0.91      0.93      0.92      5203
           1       0.96      0.94      0.95      8630

    accuracy                           0.94     13833
   macro avg       0.93      0.94      0.93     13833
weighted avg       0.94      0.94      0.94     13833



## Using Ensemble Models

### Balanced Random Forest Classifier

In [36]:
from imblearn.ensemble import BalancedRandomForestClassifier
model8 = BalancedRandomForestClassifier(random_state = 0)
model8.fit(X_train, y_train)
y_pred_rf = model8.predict(X_test) # Predictions for the X test

In [37]:
# Accuracy score and classification report

from sklearn.metrics import classification_report, accuracy_score
print(accuracy_score(y_test, y_pred_rf))
print(classification_report(y_test, y_pred_rf))

0.7298850574712644
              precision    recall  f1-score   support

           0       0.99      0.71      0.83     10041
           1       0.22      0.90      0.36       921

    accuracy                           0.73     10962
   macro avg       0.61      0.81      0.59     10962
weighted avg       0.92      0.73      0.79     10962



## Using XG Boost Classifier

In [38]:
from xgboost import XGBClassifier
model9 = XGBClassifier(scale_pos_weight = 10)
model9.fit(X_train, y_train)
y_pred_xgb = model9.predict(X_test) # Predictions for the X test

C:\Users\cti110016\Anaconda3\lib\site-packages\xgboost\sklearn.py:1224: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[17:47:11] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.5.1/src/learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


In [39]:
# Accuracy score and classification report

from sklearn.metrics import classification_report, accuracy_score
print(accuracy_score(y_test, y_pred_xgb))
print(classification_report(y_test, y_pred_xgb))

0.8375296478744755
              precision    recall  f1-score   support

           0       0.97      0.85      0.91     10041
           1       0.30      0.72      0.43       921

    accuracy                           0.84     10962
   macro avg       0.64      0.78      0.67     10962
weighted avg       0.91      0.84      0.87     10962



### Conclussion: 

* Without Resampling: Recall = 25%
* With Oversampling: Recall = 98%
* With undersampling: Recall = 71%

```We can see that Oversampling Performs better than Undersampling, as Undersampling leads to Huge Data Loss, making it Difficult for the Model to Learn the Rules Effectively.```